### PR TITLE
[NTOS:CM] Avoid unnecessary ObDereferenceObject when handling an ObInsertObject failure

### PR DESCRIPTION
**Pull request by wnstngs**
## Purpose

The problem is detailed in the JIRA ticket: https://jira.reactos.org/browse/CORE-17904. In case of failure, `ObInsertObject` itself dereferences the object, so calling `ObDereferenceObject` one more time in the failure handler is redundant.

JIRA issue: [CORE-17904](https://jira.reactos.org/browse/CORE-17904)

## Proposed changes

- Remove `ObDereferenceObject` call from `CmpCreateRegistryRoot` in case `ObInsertObject` fails.
- Add missing `RootKey` dereference if `CmpCreateRegistryRoot` failed to get the key cell (`KeyCell`).

### CODE REVIEW

**Review by HBelusca — DISMISSED** [review]
Cc @SergeGautherie too

**Review by tkreuzer — COMMENTED** [review]
The behavior is very weird, because you wouldn't expect a function to just blatently delete your object, if it failed to insert it. I assume that behavior has been copied from MS. Since this function is exported we need to replicate it's behavior. But then I would like to see a test that shows that this is actually the correct behavior. Writing such a test is also a great opportunity for learning to become a good kernel developer. :)

File: ntoskrnl/config/cmsysini.c [hunk 1]
```diff
@@ -1144,7 +1144,6 @@ CmpCreateRegistryRoot(VOID)
                             &CmpRegistryRootHandle);
     if (!NT_SUCCESS(Status))
     {
-        ObDereferenceObject(RootKey);
```
Comment (SergeGautherie) [inline ntoskrnl/config/cmsysini.c:1147]: Right. Moreover, it needs to be added at line 1112.

Comment (wnstngs) [reply ntoskrnl/config/cmsysini.c:1147]: Could you please elaborate on why? As I see ObReferenceObjectByHandle is called on line 1151, and no reference occurs before 1112

Comment (HBelusca) [reply ntoskrnl/config/cmsysini.c:1147]: @wnstngs Isn't the ObCreateObject giving the initial reference? (the object now exists after all...)

Comment (SergeGautherie) [reply ntoskrnl/config/cmsysini.c:1147]: For the same reason as old lines 1124 and 1160.
That `ObReferenceObjectByHandle()` doesn't matter for this PR.

**Review by SergeGautherie — APPROVED** [review]

**Review by ThFabba — APPROVED** [review]

**Review by HBelusca — APPROVED** [review]

**Review by GeoB99 — APPROVED** [review]
De-referencing the root key object when failing to get the hive cell is correct. What's needed is to edit the PR title to fit your changes accordingly and add `[NTOS:CM]` at the beginning of your commit title.

**Pull request merged**